### PR TITLE
Adds exception objects to log messages.

### PR DESCRIPTION
**Pull request by MaineC**
Also forwards exception objects through failure listeners eliminates another 17
calls to getMessage().

Relates to #10021

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/Base64.java [hunk 1]
```diff
@@ -161,6 +161,9 @@
  * @author rob@iharder.net
  * @version 2.3.7
  */
+
+// Given the class comment above ^^^^^ the license header on top of this file looks bogus
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/Base64.java]: should we have a followup issue for this? I don't think this is the only place we have this problem...

File: core/src/test/java/org/elasticsearch/discovery/ZenFaultDetectionTests.java [hunk 2]
```diff
@@ -192,8 +192,8 @@ public void testMasterFaultDetectionConnectOnDisconnect() throws InterruptedExce
         final CountDownLatch notified = new CountDownLatch(1);
         masterFD.addListener(new MasterFaultDetection.Listener() {
 
-            @Override
-            public void onMasterFailure(DiscoveryNode masterNode, String reason) {
+            @Override   // TODO ignores cause
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/discovery/ZenFaultDetectionTests.java]: can we create a follow up issue?

**Comment by rjernst** [comment]
This is great. These errors should be a little easier to debug now! LGTM

File: core/src/main/java/org/elasticsearch/common/Base64.java (hunk 1 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/common/Base64.java]: Created #14857 for it.

File: core/src/test/java/org/elasticsearch/discovery/ZenFaultDetectionTests.java (hunk 2 above)
Comment (MaineC) [inline core/src/test/java/org/elasticsearch/discovery/ZenFaultDetectionTests.java]: Done: #15004 (hope the wording is clear enough...)

**Pull request merged**